src/tribol/mesh/MfemData.hpp

```diff
@@ -0,0 +1,1390 @@
+   * frictionless contact.
+   */
+  static std::vector<real*> GetRedecompFieldPtrs(mfem::GridFunction& redecomp_fn);
+
+private:
+  /**
+   * @brief Creates and stores data that changes when the redecomp mesh is
+   * updated
+   */
+  struct UpdateData
+  {
+    /**
+     * @brief Construct a new UpdateData object
+     * 
+     * @param xfer Submesh to redecomp field transfer object
+     * @param submesh Grid function on the (coarse) submesh
+     */
+    UpdateData(
+      const SubmeshRedecompTransfer& xfer,
+      const mfem::ParGridFunction& submesh
+    );
+    
+    /**
+     * @brief Submesh to redecomp field transfer object
+     */
+    const SubmeshRedecompTransfer& xfer_;
+
+    /**
+     * @brief Grid function values on the redecomp mesh
+     */
+    mfem::GridFunction redecomp_;
+  };
+  
+  /**
+   * @brief Get the UpdateData object
+   * 
+   * @return UpdateData& 
+   */
+  UpdateData& GetUpdateData();
+
+  /**
+   * @brief Get the UpdateData object
+   * 
+   * @return const UpdateData& 
+   */
+  const UpdateData& GetUpdateData() const;
+
+  /**
+   * @brief Grid function on the submesh
+   *
+   * @note Stored as a reference wrapper so the reference can be updated
+   */
+  std::reference_wrapper<const mfem::ParGridFunction> submesh_;
+
+  /**
+   * @brief UpdateData object created upon call to UpdateField()
+   */
+  std::unique_ptr<UpdateData> update_data_;
+};
+
+/**
+ * @brief Stores MFEM and transfer data associated with parent vector fields
+ * (displacement and velocity)
+ */
+class MfemMeshData
+{
+public:
+  /**
+   * @brief Construct a new MfemMeshData object
+   * 
+   * @param mesh_id_1 Integer identifier for first Tribol registered mesh
+   * @param mesh_id_2 Integer identifier for second Tribol registered mesh
+   * @param mesh Volume mesh of parent domain
+   * @param current_coords Grid function (on mesh) holding current coordinates
+   * @param attributes_1 Mesh boundary attributes identifying first mesh
+   * @param attributes_2 Mesh boundary attributes identifying second mesh
+   */
+  MfemMeshData(
+    integer mesh_id_1,
+    integer mesh_id_2,
+    mfem::ParMesh& mesh,
+    const mfem::ParGridFunction& current_coords,
+    const std::set<integer>& attributes_1,
+    const std::set<integer>& attributes_2
+  );
+
+  /**
+   * @brief Get coordinate grid function on the parent mesh
+   * 
+   * @return const mfem::ParGridFunction& 
+   */
+  const mfem::ParGridFunction& GetParentCoords() const
+  {
+    return coords_.GetParentField();
+  }
+
+  /**
+   * @brief Sets a new coordinate grid function on the parent mesh
+   * 
+   * @param current_coords Coordinate grid function on the parent mesh
+   */
+  void SetParentCoords(const mfem::ParGridFunction& current_coords);
+
+  /**
+   * @brief Build a new redecomp mesh and update grid functions
+   *
+   * @note This method should be called after the coordinate grid function has
+   * changed.
+   */
+  void UpdateMeshData();
```

Comment:
When I saw this used in Tribol.cpp I think, it was not clear to me at all that this builds a redecomp mesh. Does this routine hold all of the goods to go from parent mfem volume parmesh -> submeshes -> LOR meshes (if parent is HO) -> redecomp meshes? If so, we may want to consider a function name that indicates, at least a little, that ALL of this is happening. UpdateMeshData() just sounds like you are updating existing data for existing meshes, not building an entirely new redecomposition with this workflow.

Reply:
The name of the method is meant to convey it updates the `MfemMeshData` object and the naming is meant to mimic `tribol::update()`.  I've updated it to `UpdateMfemMeshData()` to hopefully convey it's updating the object rather than data on a mesh.  The doxygen comment should let the user know what's happening here, but I also added a comment in `mfem_tribol.cpp`.